### ISSUE TITLE
 Not possible to use marshmallow schemas as part of decorations

### ISSUE THREAD

**Issue by leamingrad** [brunoais/apispec-decorated-crawler#2]
(Try again with an actual description...)

When using the plugin with the latest version of `apispec` from PyPI, it is not currently possible to use marshmallow schemas in the docstrings of the decorators.

Specifically, in the example in the documentation the `calling` decorator references the `PetSchema` schema, but the resulting YAML only resolves the schema (to `#/definitions/Pet`) for the function that has been decorated.

I've done some investigation and this appears to be due to the order in which the plugins are applied by `apispec`. Specifically, the code at https://github.com/marshmallow-code/apispec/blob/1b58e3c285a8796215f7f214bd88b3a8189e14a1/apispec/core.py#L266 iterates through the available plugins and applies the operation helpers one-by-one.

However, as `apispec-decorated-crawler` uses the old method of declaring its helper functions, it will not be called until https://github.com/marshmallow-code/apispec/blob/1b58e3c285a8796215f7f214bd88b3a8189e14a1/apispec/core.py#L272, at which point Marshmallow schema resolution has already occurred.

I'm not sure what the best solution is here, but I think the simplest would just be updating the plugin to use the new plugin API documented at https://apispec.readthedocs.io/en/stable/writing_plugins.html to declare the operation helper - then users could just put this plugin in the list before the marshmallow plugin in order to get schema resolution to work.

I am happy to send a PR with this in, but would like some guidance as to whether it is the right approach first.

**Comment by brunoais** [comment]
I actually intended to use the new API when I did this code. Unfortunately, the new plugin system was not ready by that time and I needed this for myself too.
Indeed... they really shouldn't have made a backwards breaking interface update.

Updating this code to use the new interface is the best approach to do. I just currently don't have the time to do it. If you may, that's great! I can supply help in what I can.
Do note I don't have tests due to how tight my scheduling has been, so if you can supply some manual testing notes, that would be great.

**Comment by brunoais** [comment]
@leamingrad Are you working on this? Do you have any questions?

**Comment by leamingrad** [comment]
Yeah sorry I had a very quick stab at this in December but then got swamped by other stuff so haven't had a chance to pick it back up.

I am happy to take this on, and should have some free time next weekend to push an MR. 

Quick question: I had a quick go at writing some automated tests for this - mostly just end-to-end tests that check that the examples in the documentation work. Would you be willing to accept an MR that includes these, and if so do you have any thoughts on what the best testing strategy would be?

**Comment by brunoais** [comment]
OK!

> I had a quick go at writing some automated tests for this - mostly just end-to-end tests that check that the examples in the documentation work. Would you be willing to accept an MR that includes these

Yes please. This is in a dire need for tests which have been put more down in my schedule due to lack of time (priorities).

> if so do you have any thoughts on what the best testing strategy would be?

I don't really know which one is the best. I do have some ideas for the strategy:

- Use pytest to the extent we can, including to make fixtures that mock the expected parts of apispec
- Use pytest parameterization markers for all variables and potential variables. Do not parametrize fixtures directly unless there's a good reason to.
- Use `unittest.mock` as the mocks library. Use mocks to check that the "key methods" are called with the right information (or key information)

You may focus on integration tests than unit tests (even though using `unittest` library), if you prefer.

**Comment by leamingrad** [comment]
I've had a quick go at producing the referenced MR that adds a basic end-to-end test. I'd guess it is easier to move comments on the strategy there?
